### PR TITLE
1461 verify loop type in dep tools

### LINKED ISSUE #1461: Dependency tools do not check validity of loop type
The constructor of the dependency tools does not check if the loop type to be considered for parallelization is actually valid (i.e. defined in the config file). Specifying an invalid loop type should raise an exception.

### PR DESCRIPTION
**Pull request by hiker**
Fixes #1461. I did some minor pylint fixes, but the configuration class would actually need some refactoring (1300 lines by now, and it contains three API-specific config classes and a base class - these should be moved into separate domain/XXX directories). But due to the number of changes required by that I didn't want to do that without checking with other working happening atm.

I made the access to `VALID_LOOP_TYPES` consistent across all APIs - via an attribute in the domain-specific constant class (which LFRic and GOcean already did).

### CODE REVIEW

File: src/psyclone/configuration.py [hunk 1]
```diff
@@ -692,9 +693,17 @@ def get_default_keys(self):
         '''
         return self._config.defaults()
 
+    def get_constants(self):
+        ''':returns: the constants instance of the current API.
+        :rtype: either :py:class:`psyclone.domain.lfric.LFRicConstants`, \
+            :py:class:`psyclone.domain.gocean.GOceanConstants`, or \
+            :py:class:`psyclone.domain.nemo.NemoConstants`
+            '''
```
Comment (arporter) [inline src/psyclone/configuration.py:701]: Please un-indent the closing ''' and remove the following blank line.

File: src/psyclone/configuration.py [hunk 2]
```diff
@@ -692,9 +693,17 @@ def get_default_keys(self):
...
+
+        return self.api_conf().get_constants()
+
 
 # =============================================================================
-# TODO (issue #1163): Add support for lists in the configuration file
```
Comment (arporter) [inline src/psyclone/configuration.py:697]: Well spotted.

File: src/psyclone/configuration.py [hunk 3]
```diff
@@ -1314,11 +1326,13 @@ def get_loop_type_data(self):
         return self._loop_type_data
 
     def get_valid_loop_types(self):
-        '''
-        :returns: a list of valid loop types.
+        '''The list is sorted to have reproducible results for testing.
```
Comment (arporter) [inline src/psyclone/configuration.py:1329]: Please put a new line between the opening ''' and the sentence and then please follow the sentence with a blank line.

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 4]
```diff
@@ -65,10 +68,28 @@ class DependencyTools(object):
...
     def __init__(self, loop_types_to_parallelise=None,
                  language_writer=FortranWriter()):
         if loop_types_to_parallelise:
+            # Verify that all loop types specified are valid:
+            config = Config.get()
+            constants = config.get_constants()
+            for loop_type in loop_types_to_parallelise:
+                if loop_type not in constants.VALID_LOOP_TYPES:
+                    if six.PY2:
+                        out_list = [str(l_type)for l_type in
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:84]: missing space before "for"

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good Joerg.
All tests and examples are OK with compilation.
Almost all changes are pylint/pycodestyle clean (see inline).
Please could you cast your eye over the 'Configuration' section of the Dev Guide and check whether it needs updating, e.g. do all APIs now have VALID_LOOP_TYPES?

File: src/psyclone/configuration.py (hunk 1 above)
Comment (hiker) [reply src/psyclone/configuration.py]: Done (my pylint/pycodestyle did not complain about this *sigh*)

File: src/psyclone/configuration.py (hunk 3 above)
Comment (hiker) [reply src/psyclone/configuration.py]: Done.

File: src/psyclone/psyir/tools/dependency_tools.py (hunk 4 above)
Comment (hiker) [reply src/psyclone/psyir/tools/dependency_tools.py]: Strangely again my pylint did not detect this :(

**Comment by hiker** [comment]
>Please could you cast your eye over the 'Configuration' section of the Dev Guide and check whether it needs updating, e.g. do all APIs now have VALID_LOOP_TYPES?

Good catch, I missed that. I've reworded that section and added VALID_LOOP_TYPES.

Ready for next review.

File: src/psyclone/psyir/tools/dependency_tools.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/psyir/tools/dependency_tools.py]: I dub myself "Andy Pylint Porter" :-)

**Comment by arporter** [comment]
All requested changes have been made. Docs have been updated and build cleanly.
All tests and examples are OK with compilation.
Issue number is not mentioned anywhere.

**Review by arporter — APPROVED** [review]

**Pull request merged**